### PR TITLE
feat(focus-trap): allow setting initially focused element

### LINKED ISSUE #2224: md-sidenav mode="over" focus state
#### Bug, feature request, or proposal:
The base path remains in active state when you select another link and you close and reopen the sidenav, only in mode="over". Need an event, like a click, to deactivate it.

#### What is the expected behavior?
It has worked fine up to 2.0.0-alpha.10

#### What are the steps to reproduce?

http://plnkr.co/edit/R9p0F7

#### Which versions of Angular, Material, OS, browsers are affected?
Material 2.0.0-alpha.11-xx, in all browsers.

#### Is there anything else we should know?
I tried Angular workarounds (routerLinkActive) without success.

Thanks

### PR DESCRIPTION
**Pull request by mmalerba**
also fix `md-nav-list` focused item style.

fixes #2224

### CODE REVIEW

File: src/lib/core/a11y/focus-trap.ts [hunk 1]
```diff
@@ -97,27 +101,47 @@ export class FocusTrap {
     this._ngZone.onMicrotaskEmpty.first().subscribe(() => this.focusLastTabbableElement());
   }
 
+  private _getRegionMarker(marker: 'start' | 'end'): HTMLElement | null {
```
Comment (jelbourn) [inline src/lib/core/a11y/focus-trap.ts]: Add function description? Not immediately clear what a "region marker" is.

File: src/lib/list/list.ts [hunk 2]
```diff
@@ -151,6 +149,7 @@ export class MdListItem implements AfterContentInit {
 
   constructor(private _renderer: Renderer2,
               private _element: ElementRef,
+              private _cdr: ChangeDetectorRef,
```
Comment (jelbourn) [inline src/lib/list/list.ts]: Where is this used?

File: src/lib/core/a11y/focus-trap.ts (hunk 1 above)
Comment (jelbourn) [inline src/lib/core/a11y/focus-trap.ts]: Instead of "marker", maybe "boundary element"?

File: src/lib/core/a11y/focus-trap.ts [hunk 3]
```diff
@@ -97,27 +101,47 @@ export class FocusTrap {
...
+    if (marker == 'start') {
+      return markers.length ? markers[0] : this._getFirstTabbableElement(this._element);
+    }
+    return markers.length ?
+        markers[markers.length - 1] : this._getLastTabbableElement(this._element);
+  }
+
+  /** Focuses the element that should be focused when the focus trap is initialized. */
+  focusInitialElement() {
+    let redirectToElement = this._element.querySelector('[cdk-focus-init]') as HTMLElement;
```
Comment (jelbourn) [inline src/lib/core/a11y/focus-trap.ts]: Write out `initial` (to disambiguate from "initialize")?

File: src/lib/core/a11y/focus-trap.spec.ts [hunk 4]
```diff
@@ -167,8 +173,9 @@ class FocusTrapWithBindings {
   template: `
     <div cdkTrapFocus>
       <input>
-      <button id="last" cdk-focus-end></button>
-      <button id="first" cdk-focus-start>SAVE</button>
+      <button id="last" cdk-focus-region-end></button>
+      <button id="middle" cdk-focus-init></button>
+      <button id="first" cdk-focus-region-start>SAVE</button>
```
Comment (jelbourn) [inline src/lib/core/a11y/focus-trap.spec.ts]: Why does start come after end here?

File: src/lib/core/a11y/focus-trap.ts [hunk 5]
```diff
@@ -97,27 +101,47 @@ export class FocusTrap {
     this._ngZone.onMicrotaskEmpty.first().subscribe(() => this.focusLastTabbableElement());
   }
 
+  private _getRegionMarker(marker: 'start' | 'end'): HTMLElement | null {
+    let markers = [
+      ...Array.prototype.slice.call(this._element.querySelectorAll(`[cdk-focus-region-${marker}]`)),
+      ...Array.prototype.slice.call(this._element.querySelectorAll(`[cdk-focus-${marker}]`)),
```
Comment (jelbourn) [inline src/lib/core/a11y/focus-trap.ts]: Add comment explaining what's for backwards compatibility

File: src/lib/list/list.ts (hunk 2 above)
Comment (mmalerba) [inline src/lib/list/list.ts]: nowhere, switched to adding class via renderer

File: src/lib/core/a11y/focus-trap.spec.ts (hunk 4 above)
Comment (mmalerba) [inline src/lib/core/a11y/focus-trap.spec.ts]: The test was already like this, I think the reasoning is to make sure that it's really grabbing the element with the attribute, not just the first or last one. I changed it to something less confusing.

**Comment by mmalerba** [comment]
comments addressed, PTAL

**Pull request merged**